pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -0,0 +1,714 @@
+		return nil, -1, nil
+	}
+	listOption := &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: selector,
+	}
+	err = r.List(context.TODO(), rsList, listOption)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	rsSet := make(map[types.UID]struct{})
+	for _, rs := range rsList.Items {
+		controllerRef := metav1.GetControllerOf(&rs)
+		if controllerRef.UID == deployment.UID {
+			rsSet[rs.UID] = struct{}{}
+		}
+	}
+
+	// List all Pods owned by this Deployment.
+	podList := &corev1.PodList{}
+	if err := r.List(context.TODO(), podList, listOption); err != nil {
+		return nil, -1, err
+	}
+
+	matchedPods := make([]*corev1.Pod, 0, len(podList.Items))
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		controllerRef := metav1.GetControllerOf(pod)
+		if controllerRef == nil {
+			continue
+		}
+		// filter UID
+		if _, ok := rsSet[controllerRef.UID]; ok {
+			matchedPods = append(matchedPods, pod)
+		}
+	}
+	return matchedPods, *deployment.Spec.Replicas, nil
+}
+
+func (r *ReconcileWorkloadSpread) getPodReplicasSet(ref *appsv1alpha1.TargetReference, namespace string) (*scaleAndSelector, error) {
+	ok, _ := wsutil.VerifyGroupKind(ref, controllerKindRS.Kind, []string{controllerKindRS.Group})
+	if !ok {
+		return nil, nil
+	}
+
+	rs := &appsv1.ReplicaSet{}
+	err := r.Get(context.TODO(), client.ObjectKey{Namespace: namespace, Name: ref.Name}, rs)
+	if err != nil {
+		// when error is NotFound, it is ok here.
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return &scaleAndSelector{
+		UID:      rs.UID,
+		scale:    *(rs.Spec.Replicas),
+		selector: rs.Spec.Selector,
+	}, nil
+}
+
+func (r *ReconcileWorkloadSpread) getPodJob(ref *appsv1alpha1.TargetReference, namespace string) (*scaleAndSelector, error) {
+	ok, _ := wsutil.VerifyGroupKind(ref, controllerKindJob.Kind, []string{controllerKindJob.Group})
+	if !ok {
+		return nil, nil
+	}
+
+	job := &batchv1.Job{}
+	err := r.Get(context.TODO(), client.ObjectKey{Namespace: namespace, Name: ref.Name}, job)
+	if err != nil {
+		// when error is NotFound, it is ok here.
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return &scaleAndSelector{
+		UID:      job.UID,
+		scale:    *(job.Spec.Parallelism),
+		selector: job.Spec.Selector,
+	}, nil
+}
+
+// getPodsForWorkloadSpread returns Pods managed by the WorkloadSpread object.
+// return two parameters
+// 1. podList for workloadSpread
+// 2. workloadReplicas
+func (r *ReconcileWorkloadSpread) getPodsForWorkloadSpread(ws *appsv1alpha1.WorkloadSpread) ([]*corev1.Pod, int32, error) {
+	var scaleSelector *scaleAndSelector
+	var err error
+
+	if ws.Spec.TargetReference != nil {
+		// Deployment manages Pods by ReplicaSet, so it has a different method.
+		if ws.Spec.TargetReference.Kind == controllerKindDep.Kind {
+			return r.getPodDeployment(ws.Spec.TargetReference, ws.Namespace)
+		}
+
+		for _, finder := range r.finders() {
```

Comment:
Why not just a `switch case` here to match the exact gvk and it's get function? The `finders` comes from `PDB` or `PUB` controller is because they can define a selector to match multiple workloads. But `WorkloadSpread` can just define the target reference to a definite workload.

After changed to `switch case`, there is no need to `VerifyGroupKind` in each get function of workloads.

Reply:
Okay, I will fix it.